use Google's published layoutlib

gabrielittner opened the pull request:
Closes #1221.
Closes #884.
Closes #717.

- Replaces the manually published layoutlib modules with the new artifacts published by Google.
- In 14.0.3 a new `layoutlib-resources` artifact was introduced. This contains the icu, keyboard and font files that previously were in the main layoutlib artifacts. It also contains a `res` folder which can be used instead of reading resources from the SDK platform directory.
- In 14.0.4 icu, keyboard and font files moved back to the main layoutlib artifacts. `layoutlib-resources` now only contains the `res` folder which moved to a top level directory inside the jar.
- ~Initially I've also upgraded `layoutlib-api` to the Jellyfish preview versions. However it's not actually needed for everything to work, so I downgraded it again. Because of that this PR does not depend on any alpha/beta versions, but I guess the `14.0.x` layoutlib artifacts could still be considered as not stable until AS Jellyfish is stable.~

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/internal/resources/base/BasicResourceItem.kt:
```diff
@@ -91,8 +91,6 @@ abstract class BasicResourceItem(
     }
   }
 
-  override fun setValue(value: String?): Unit = throw UnsupportedOperationException()
```
Method was removed from the base class/interface

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/internal/Renderer.kt:
```diff
@@ -114,7 +114,7 @@ internal class Renderer(
     val fontLocation = File(platformDataDir, "fonts")
     val nativeLibLocation = File(platformDataDir, getNativeLibDir())
-    val icuLocation = File(platformDataDir, "icu" + File.separator + "icudt70l.dat")
+    val icuLocation = File(platformDataDir, "icu" + File.separator + "icudt72l.dat")
```
File name changed in the new layoutlib version

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:
```diff
@@ -616,7 +607,7 @@ class Paparazzi @JvmOverloads constructor(
       // https://android.googlesource.com/platform/frameworks/layoutlib/+/d58aa4703369e109b24419548f38b422d5a44738/bridge/src/com/android/layoutlib/bridge/BridgeRenderSession.java#171
       // BridgeRenderSession.executeCallbacks aggressively tears down the main Looper and BridgeContext, so we call the static delegates ourselves.
-      Handler_Delegate.executeCallbacks()
+      Handler_Delegate.executeCallbacks(System_Delegate.nanoTime())
```
This is effectifely what `executeCallbacks` used to do internally before, but now it requires the time to be passed in

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/Paparazzi.kt:
```diff
@@ -352,29 +353,19 @@ class Paparazzi @JvmOverloads constructor(
+    // System_Delegate.nanoTime() is nanoTime - previousNanoTime + sChoreographerTime.
+    // Because of that setNanosTime needs to be called with 0 first so that the previous time will be 0 for the actual
+    // call. sChoreographerTime does not matter because doFrame which modifies it is not called by Paparazzi.
+    System_Delegate.setNanosTime(0L)
     // Execute the block at the requested time.
     System_Delegate.setNanosTime(frameNanos)
 
-    val choreographer = Choreographer.getInstance()
-    val areCallbacksRunningField = choreographer::class.java.getDeclaredField("mCallbacksRunning")
-    areCallbacksRunningField.isAccessible = true
+    // should happen before doCallbacks so that compose re-composition happens
+    executeHandlerCallbacks()
+    // 1 is the only callbackType allowed by doCallbacks
+    Choreographer_Delegate.doCallbacks(Choreographer.getInstance(), 1, frameNanos)
```
`Choreographer_Delegate` also has a `doFrame` method that updates the choreographer time and runs the callbacks. But I decided against using it for now because `sChoreographerTime` is only updated after the callbacks run which would require changing some tests that rely on the time to be already updated and I wanted to avoid updating the tests for now. It also a bit more complicated to use because `sChoreographerTime` can apparently not become smaller without breaking things (`PaparazziTest.animationCallbacksForStaticSnapshots` was suceeding when run on its own but failed when running all of `PaparazziTest` when I tried using doFrame and starting with `0` in each test).

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/DeviceConfig.kt:
```diff
@@ -261,7 +261,7 @@ data class DeviceConfig(
       ydpi = 440,
       orientation = ScreenOrientation.PORTRAIT,
-      density = Density.DPI_420,
+      density = Density.create(420),
```
Density was changed from an enum to a class.

TWiStErRob replied:
Hmm, someone was asking for something like this.

TWiStErRob commented on gradle/libs.versions.toml:
```diff
@@ -11,11 +11,9 @@ ktlint = "0.50.0"
+compileSdk = "34"
 
-# Maps to this commit: https://android.googlesource.com/platform/prebuilts/studio/layoutlib/+/5e972ea
-layoutlib = "2022.3.1-5e972ea"
-layoutlibPrebuiltSha = "5e972ea"
+layoutlib = "14.0.1"
```
How much of this PR is actually part of #1216 and #1220?

gabrielittner replied:
The compileSdkVersion bump can and probably needs to already happen with the Iguana update. As for the API changes, I'd need to try building it to find out.

gabrielittner replied:
I've rebased this PR and restructured the commits. The first commit bumps the version to Iguana 2023.2.1-beta02-6c7316c, the following commits then apply the required changes for Iguana and the last commit bumps to Jellyfish and Google's artifacts. All the changes here except for that last commit are already needed to make the Iguana layoutlib run.

Iguana has the same issue with the 3 tests that are failing on either macOS or on Windows+Linux, depending on where the snapshot was created. I made a test run with a manually published 2023.2.1-beta02-6c7316c here https://github.com/freeletics/paparazzi/actions/runs/7590152036

gabrielittner commented on paparazzi/src/main/java/app/cash/paparazzi/internal/resources/base/BasicFileResourceItem.kt:
```diff
@@ -86,7 +86,7 @@ open class BasicFileResourceItem(
         return BasicFileResourceItem(resourceType, name, configuration, visibility, relativePath)
       }
-      val density = Density.values()[encodedDensity - 1]
+      val density = Density.create(encodedDensity)
```
https://cs.android.com/android-studio/platform/tools/base/+/a728d0c33ce30df27187501c61fdbe6e760342af:resource-repository/main/java/com/android/resources/base/BasicFileResourceItem.java;dlc=471934f339e1ccc75ab79d8def98ff876f1eb75f

gabrielittner commented on .github/workflows/build.yml:
```diff
@@ -18,7 +18,7 @@ jobs:
       matrix:
         os: [
-          macOS-latest,
+          macos-14,
```
This uses the newer ARM based macOS runner and makes `composeA11y`, `customFontsInCode`, and `customFontsInXml`. The 3 tests would still fail on Intel macOS, it seems that that variant of layoutlib has a small but bigh enough rendering difference compared to macOS ARM, Linux and Windows. 

I've also did a test with the latest tagged Iguana layoutlib by publishing it myself and with the same changes as this PR (except for the last commit [3b2c9c4](https://github.com/cashapp/paparazzi/pull/1228/commits/3b2c9c4da4d28cd172a796991f9c3d2152513d13)) Iguana also works: https://github.com/freeletics/paparazzi/actions/runs/7731866563

gabrielittner commented on paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt:
```diff
@@ -76,6 +76,7 @@ public class PaparazziPlugin : Plugin<Project> {
   private fun <T> setupPaparazzi(project: Project, variants: DomainObjectSet<T>) where T : BaseVariant, T : TestedVariant {
     project.addTestDependency()
     val nativePlatformFileCollection = project.setupNativePlatformDependency()
+    val layoutlibResourcesFileCollection = project.setupLayoutlibResourcesDependency()
```
Everything except for the native lib itself moved to a new `layoutlib-resources` artifact in 14.0.3, so we need to unzip that one as well and have 2 system properties for passing the directories to the test

gabrielittner commented:
With the Iguana update merged I've rebased this and updated the PR description. Other than the new `layoutlib-resources` artifact that reduces the reliance on the local Android SDK directory there seems to be no required change to the Paparazzi code from this update.

TWiStErRob commented on gradle/libs.versions.toml:
```diff
@@ -14,9 +14,8 @@ moshi = "1.15.1"
-# Maps to this commit: https://android.googlesource.com/platform/prebuilts/studio/layoutlib/+/6c7316c
-layoutlib = "2023.2.1-6c7316c"
-layoutlibPrebuiltSha = "6c7316c"
+layoutlibApi = "31.3.1"
```
Should this be replaced with versions.androidTools instead?

gabrielittner replied:
I've kept it separate like before because there might still be situations in the future where this might need to be updated independently.

TWiStErRob commented on libs/layoutlib/build.gradle:
```diff
@@ -1,61 +0,0 @@
-import org.ajoberstar.grgit.Grgit
```
Only used here, there are traces of it left in toml and build-logic.

Also "closes #884"

gabrielittner replied:
Removed the leftovers 👍 [e62b010](https://github.com/cashapp/paparazzi/pull/1228/commits/e62b010f7a167a8d923f259a3b89fbcccfac35e6)

TWiStErRob replied:
Nice, also just realised, transitively "fixes #717"

gabrielittner commented:
I've updated the PR to layoutlib-api 31.4.0 and layoutlib 14.0.7 which should match the stable Jellyfish release.

The last commit updates screenshots for the following 4 tests, in all of them the compose font padding changes. I believe there might be a build cache issue that causes them not to fail on master because compose 1.6.0 changed the default font padding and the last commit on master bumped compose from 1.5.x to 1.6.x. In this pr the 4 tests started failing after just merging master into this branch.
```
PaparazziPluginTest.compose
PaparazziPluginTest.composeA11y
PaparazziPluginTest.composeWear
PaparazziPluginTest.nightModeCompose
```

gabrielittner commented:
@jrodbx With 1.3.4 out, can this get merged? Happy to make any additional changes if needed but all tests are succeeding without any changes to them

jrodbx commented on gradle/libs.versions.toml:
```diff
@@ -13,9 +13,8 @@ moshi = "1.15.1"
-# Maps to this commit: https://android.googlesource.com/platform/prebuilts/studio/layoutlib/+/6c7316c
-layoutlib = "2023.2.1-6c7316c"
-layoutlibPrebuiltSha = "6c7316c"
+layoutlibApi = "31.4.1"
```
inline, since only one use

jrodbx commented on gradle/libs.versions.toml:
```diff
@@ -13,9 +13,8 @@ moshi = "1.15.1"
-layoutlib = "2023.2.1-6c7316c"
-layoutlibPrebuiltSha = "6c7316c"
+layoutlibApi = "31.4.1"
+layoutlib = "14.0.8"
```
move up to line 11, since it's now just another "normal" version

jrodbx commented on gradle/libs.versions.toml:
```diff
@@ -57,7 +50,10 @@ moshi-kotlinReflect = { module = "com.squareup.moshi:moshi-kotlin", version.ref
 tools-common = { module = "com.android.tools:common", version.ref = "androidTools" }
-tools-layoutlib = { module = "com.android.tools.layoutlib:layoutlib-api", version = "31.4.1" }
+tools-layoutlib-api = { module = "com.android.tools.layoutlib:layoutlib-api", version.ref = "layoutlibApi" }
+tools-layoutlib = { module = "com.android.tools.layoutlib:layoutlib", version.ref = "layoutlib" }
```
nit: ordering, move this one line up (above api)

jrodbx commented on gradle/libs.versions.toml:
```diff
@@ -57,7 +50,10 @@ moshi-kotlinReflect = { module = "com.squareup.moshi:moshi-kotlin", version.ref
+tools-layoutlib-api = { module = "com.android.tools.layoutlib:layoutlib-api", version.ref = "layoutlibApi" }
+tools-layoutlib = { module = "com.android.tools.layoutlib:layoutlib", version.ref = "layoutlib" }
+tools-layoutlib-runtime = { module = "com.android.tools.layoutlib:layoutlib-runtime", version.ref = "layoutlib" }
+tools-layoutlib-resources = { module = "com.android.tools.layoutlib:layoutlib-resources", version.ref = "layoutlib" }
```
nit: ordering, move this one line up (above runtime)

jrodbx reviewed (approved):
nice work!

jrodbx commented:
actually, i'll just merge this in now, and address nits in a followup.

yogurtearl commented:
Swapping transitive deps might warrant a bigger version bump? i.e. `1.4.0` rather than `1.3.5`?
